ontology/obo/tree.go

```diff
@@ -0,0 +1,160 @@
+	_, err = fmt.Fprintf(out, "\"%s\" [label = \"%s\"];\n", term.Id, term.Name)
+	exception.PanicOnErr(err)
+
+	// now we recursively visit children
+	for _, child := range term.Children {
+		_, err = fmt.Fprintf(out, "\"%s\" -> \"%s\";\n", term.Id, child.Id)
+		termToDot(*child, out, visitedNode)
+	}
+}
+
+// ToDot formats an input Obo slice as a tree and writes the resulting tree to a file in DOT format.
+func ToDot(outFile string, terms []Obo) {
+	var err error
+	out := fileio.EasyCreate(outFile)
+	BuildTree(terms)
+
+	//write header line for Dot
+	_, err = fmt.Fprintf(out, "digraph G{\n")
+	exception.PanicOnErr(err)
+
+	var visitedNode = make(map[string]bool)
+
+	for _, term := range terms {
+		termToDot(term, out, visitedNode)
+	}
+
+	//write tail line for Dot and close file
+	_, err = fmt.Fprintf(out, "}\n")
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+// makeTermMap creates a map to find nodes based on their ID and AltIDs
+func makeTermMap(terms []Obo) map[string]*Obo {
+	var altId string
+	termMap := make(map[string]*Obo)
+	for i := range terms {
+		term := &terms[i] // get a memory copy
+		if termMap[term.Id] != nil {
+			log.Fatalf("Term \"%v\" is duplicated in input slice of Obo.\n", term.Id)
+		}
+		termMap[term.Id] = term
+		for _, altId = range term.AltIds {
+			if termMap[altId] != nil {
+				log.Fatalf("Term \"%v\" is duplicated in input slice of Obo.\n", term.Id)
+			}
+			termMap[altId] = term
+		}
+	}
+
+	return termMap
+}
+
+// SubtreeToDot makes a DOT format tree for the subtree rooted at the input nodeId.
+// BuildTree must have been run on this tree already, as termMap is an argument.
+func SubtreeToDot(outFile string, nodeId string, termMap map[string]*Obo) {
+	var err error
+	out := fileio.EasyCreate(outFile)
+	//write header line for Dot
+	_, err = fmt.Fprintf(out, "digraph G {\n")
+	exception.PanicOnErr(err)
+
+	var visitedNode = make(map[string]bool)
+	if _, ok := termMap[nodeId]; !ok {
+		log.Fatalf("Input term not found in Obo file. Term: %v\n", nodeId)
+	}
+	termToDot(*termMap[nodeId], out, visitedNode)
+
+	//write tail line for Dot and close file
+	_, err = fmt.Fprintf(out, "}\n")
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+// NumberOfDescendents edits the field SubTreeSize of input Obo structs
+// to the number of descendent nodes in a subtree rooted on a particular node.
+func NumberOfDescendents(termMap map[string]*Obo) {
+	var visitedNode = make(map[string]bool)
+	for _, i := range termMap {
+		numberOfDescendentsRecursive(i, visitedNode)
+	}
+}
+
+// numberOfDescendentsRecursive is a helper function of NumberOfDescendents, which
+// calculates the size of subtrees rooted on a particular Obo struct in a larger
+// Obo tree.
+func numberOfDescendentsRecursive(term *Obo, visitedNode map[string]bool) {
+	if _, foundInMap := visitedNode[term.Id]; foundInMap {
+		return
+	}
+	visitedNode[term.Id] = true
+	//leaf case
+	if len(term.Children) == 0 {
+		term.SubTreeSize = 0
```

Comment:
I think we want to adjust counting or vocabulary so that I leaf is a subtree of size one, or has descendents of zero.  Right now we have a subtree size of 0.  I think the concept of a subtree is more common in math/cs, so I would vote for changing the counting.

Reply:
next commit resolves